### PR TITLE
refactor(intervals): conslidate interval conversion under `_make_interval` base compiler implementation

### PR DESCRIPTION
**Pull request by cpcloud**
Follow-up to #9799 to consolidate some repetitive interval-handling code.

### CODE REVIEW

**Comment by cpcloud** [comment]
Yep, just working through sqlite now, it's quite ... [unique](https://sqlite.org/lang_datefunc.html).

**Comment by cpcloud** [comment]
Hell yeah, it's just a sea of unfailure!

image

**Comment by cpcloud** [comment]
Ah, for postgres we're not calling `visit_Cast` ... fixing that now.

File: ibis/backends/sql/compilers/sqlite.py [hunk 1]
```diff
@@ -334,17 +331,46 @@ def visit_TimestampTruncate(self, op, *, arg, unit):
...
+
+        if floor:
+            args.append("floor")
 
-    visit_DateAdd = visit_DateSub = visit_DateArithmetic
+        if isinstance(op, (ops.TimestampAdd, ops.TimestampSub)):
+            func = self.f.datetime
+            args.append("subsec")
+        else:
+            func = self.f.date
```
Comment (cpcloud) [inline ibis/backends/sql/compilers/sqlite.py:367]: This entire method became the absolute worst.

Comment (gforsyth) [reply ibis/backends/sql/compilers/sqlite.py:367]: Yep, that's pretty awful.  But hey, at least it's adding support for 4 different targets!

**Comment by cpcloud** [comment]
Gonna do a rebase merge on this one so the sqlite `feat` gets picked up in the release notes.

**Review by gforsyth — APPROVED** [review]
nice!

**Comment by cpcloud** [comment]
Looks like I might be hitting a sqlite version issue.

**Comment by cpcloud** [comment]
sad

image

**Comment by cpcloud** [comment]
I love that if you pass a bogus argument to sqlite, it returns `NULL`.

**Comment by cpcloud** [comment]
I think I might just raise an exception for sqlite < 3.46. Otherwise we're looking at implementing this function

https://github.com/sqlite/sqlite/blob/a0789fcc4917af0cdf343c073152045a990a7313/src/date.c#L299

**Comment by cpcloud** [comment]
yyyyy

image

**Comment by cpcloud** [comment]
GHA is running on 3.40.x 😢

**Comment by cpcloud** [comment]
At this point I'm ready to declare versioning bankruptcy and `xfail` with `strict=False` since we're running these tests on the nix jobs which have the latest version of sqlite.

**Comment by cpcloud** [comment]
Oh wait I forgot we allow `condition` in backend markers. Trying that now.

**Comment by cpcloud** [comment]
Clouds look good:

```
…/ibis on  consolidate-interval-implementations is 📦 v9.3.0 via 🐍 v3.10.14 via ❄️   impure (ibis-3.10.14-env)
❯ pytest -m snowflake -n 8 --dist loadgroup --snapshot-update -q && pytest -m bigquery -n auto --dist loadgroup --snapshot-update -q
bringing up nodes...
................................................xx.........................................x................................................................................x......... [  9%]
.........x......................x........s................x.....x.x..................x................x.....x..............x...xx..x.....x......xxx.x..x.....................x..xxx.xx [ 19%]
.......x.......x.s..............x.x.xx.xxx.x..xxx.x.xxx..x.x..xx.x.x...x..xx..xx..x.x.....x.xxxxxx.x..xx..x...xx.x..x.x.xx..x..x..xx.xx.xxx.x.xxxx.x.xx.....x.....xx..x............... [ 29%]
...x.x.....x......x...........................x.................xx.............x.....x...........xx..........x.............................x..xx..................x..........xx....... [ 39%]
...........x.......x..............x....................................................x....x..........x.x............................................................................ [ 49%]
........................................xx.....x...........x.x......................xx......x..x...........x....x.........x...x.............x...........x.x....x.......x.............. [ 59%]
....x...x..x.x............x......x.x.x.x..x.x..x..xxxxx.xxxxx.....xx..x...x................x.....................s......................s..s.........................xx.xx.xxxx.x.x.xx [ 68%]
....x.....x...........x........x..............................x.............x......x...........x.x...x...x....x......x...................x..........x..........................x...... [ 78%]
...xx...x......x..............x..........x.x.........xx......................x......................x..........x.x.................x.x...x....x.x...x...s.................x........... [ 88%]
.x........s.....................................................................s..................................................................................................... [ 98%]
......s....s..................                                                                                                                                                         [100%]
1624 passed, 10 skipped, 216 xfailed in 210.48s (0:03:30)
bringing up nodes...
.....................................................................................................................................................................x...sssssssssssss [  8%]
ssssssss...xx.................sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssx....x.........x...........x.xx...........ss.......xx.........x..xx........x......x...x.. [ 17%]
..x.s.................x.x......x.xxx....x..xx..xx.x.xxxxxxxxxxxxxxxxxxxx.xx.xx.xxxxxxxx.xxxxxxxxx.xx.xxxxxxx.xxxxxxxxxxxxx.xxxxxxxxxxxxxxxx.xxxxxxxxx.xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx [ 26%]
xxxxxxxxx.xxx..x....x......x.xx...x.....s...........x...x.............x..x.............................x..x............xxx...................................x.x....x..x........xx.... [ 35%]
.xxx.x.....x..x.xx..........xx..xx.x....x..x.....x.x....x.........x...x........x..x....xx....x..........x.....x.x..x.xx.......x..xx....xx.....x.................x......x...x...x...... [ 43%]
x...x....xx..x.xxx......x...........x..xxx.....x....s..................x..........x...................x.....X.x...x.x...xx.xx.xx..xXx.....x.....x.x.x..........x...........x.....x.... [ 52%]
.x.....x.x.x.................x...x..........x.....x.x..x.x.X.x.......XxX..x...x..........xx..............x.....................x.x......x.x....x...xx...x....x....x..xx...........x... [ 61%]
x.........x...................................x............x....x......x..........x..x.......s.......x.............x............s..................................................... [ 70%]
......x..........xs...................x......x......x...x.....x.........x........x.x..x..x.x......x..x.................x......xx...x.....................x....x...........x....xxxxxxx [ 79%]
xx.....x............xx.x....x............................................................................x.........................................................x......x........... [ 87%]
......................................................................................................................x......x..........................................x............. [ 96%]
.....................................................................                                                                                                                                              [100%]
1632 passed, 92 skipped, 342 xfailed, 5 xpassed in 309.60s (0:05:09)
```

**Pull request merged**